Userland/Libraries/LibCrypto/Authentication/GHash.cpp

```diff
@@ -97,12 +97,12 @@ void galois_multiply(u32 (&z)[4], const u32 (&_x)[4], const u32 (&_y)[4])
     __builtin_memset(z, 0, sizeof(z));
 
     for (ssize_t i = 127; i > -1; --i) {
-        if ((y[3 - (i / 32)] >> (i % 32)) & 1) {
-            z[0] ^= x[0];
-            z[1] ^= x[1];
-            z[2] ^= x[2];
-            z[3] ^= x[3];
-        }
+
+        u32 branchless_condition = (y[3 - (i / 32)] >> (i % 32)) & 1;
+        z[0] ^= x[0] * branchless_condition;
+        z[1] ^= x[1] * branchless_condition;
+        z[2] ^= x[2] * branchless_condition;
+        z[3] ^= x[3] * branchless_condition;
```

Comment:
I wonder if something like this would be even faster, since it replaces the 4 `imul`s with 1 `cmovns` and 4 `and`s:
```suggestion
        u32 branchless_condition = (y[3 - (i / 32)] >> (i % 32)) & 1 ? 0xFFFFFFFF : 0x0;
        z[0] ^= x[0] & branchless_condition;
        z[1] ^= x[1] & branchless_condition;
        z[2] ^= x[2] & branchless_condition;
        z[3] ^= x[3] & branchless_condition;
```

Reply:
(`imul` has an average reciprocal throughput of 4 cycles while `cmov` and `and` have 1/3 cycles, but it might be neglible since real cores will probably optimize the special case of `1`)

Reply:
`The ?: operator adds a branch and makes this a little harder to read... However, the compiler sees right through it in either case`
Not sure what you're referring to, it doesn't add a branch.

Reply:
It adds a conceptual branch, but like I said, the branch is seen through and optimized away. It does result in slightly different output though. Links above.

Reply:
I mean sure, but I dont think there's any situation in which a compiler isn't able to convert a conditional expression to a conditional move.
(that's point of using that here instead of an if statement)